### ISSUE TITLE
Publishing hosted Blazor WASM project breaks if not self-contained

### ISSUE THREAD

**Issue by yugabe** [dotnet/runtime#75847]

**Comment by mkArtakMSFT** [comment]
@lewing can you please look into this? We don't recall any work that we did in this area to cause this, so maybe something has changed in your area or even in the SDK?

**Comment by lewing** [comment]
this isn't related to the wasm part of the build

**Comment by yugabe** [comment]
@lewing Even though the error message comes from the WASM client project? It's true that the Server project is the one getting built and published, but indirectly it both builds and publishes the (WASM) Client too, and the error message originates from there.

**Comment by lewing** [comment]
@yugabe yes, it has nothing to do with the wasm runtime. The error is coming from the illink targets and the surrounding build logic comes from the sdk and the blazor team's code in the sdk.  This doesn't look like it has anything to do with the wasm runtime directly.

cc @sbomer @vitek-karas @mkArtakMSFT @javiercn

**Comment by yugabe** [comment]
Sorry for bumping, but I just wanted to confirm this regression still stands as of .NET 7 RC2. So, not able to publish a hosted Blazor app that is framework dependent.

**What's weird is that the build succeeds in Visual Studio when using a folder publish profile.** The profile contains nothing that seems to indicate a difference in how the build works, but I'll include it here for reference.

```xml
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project>
  <PropertyGroup>
    <DeleteExistingFiles>false</DeleteExistingFiles>
    <ExcludeApp_Data>false</ExcludeApp_Data>
    <LaunchSiteAfterPublish>true</LaunchSiteAfterPublish>
    <LastUsedBuildConfiguration>Release</LastUsedBuildConfiguration>
    <LastUsedPlatform>Any CPU</LastUsedPlatform>
    <PublishProvider>FileSystem</PublishProvider>
    <PublishUrl>bin\Release\net7.0\publish\</PublishUrl>
    <WebPublishMethod>FileSystem</WebPublishMethod>
    <_TargetId>Folder</_TargetId>
    <SiteUrlToLaunchAfterPublish />
    <TargetFramework>net7.0</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <ProjectGuid>09041cbc-0fbc-4abb-8fc4-280cf26eafeb</ProjectGuid>
    <SelfContained>false</SelfContained>
  </PropertyGroup>
</Project>
```

The error message is different now in the CLI than it originally was in RC1. The message when running `dotnet publish ./BlazorWasmHosted --runtime linux-x64 --no-self-contained` became:
```
C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\7.0.0-rc.2.22472.3\Sdk\WasmApp.targets(208,5): erro
r : $(MicrosoftNetCoreAppRuntimePackDir)='', and cannot find %(ResolvedRuntimePack.PackageDirectory)=. One of these nee
d to be set to a valid path [{...}\BlazorWasmHosted\Client\BlazorWasmHosted.Client.csproj]
```

**Comment by eerhardt** [comment]
The original error is happening because the `SelfContained=false` MSBuild property is flowing into the `Client` project. This value can't be set for the Client project, because it is a WASM app. It HAS to be SelfContained.

One workaround I discovered is to modify the `Client.csproj` with:

```xml
  <PropertyGroup>
    <StaticWebAssetsAdditionalPublishPropertiesToRemove>$(StaticWebAssetsAdditionalPublishPropertiesToRemove);SelfContained</StaticWebAssetsAdditionalPublishPropertiesToRemove>
  </PropertyGroup>
```

And then when you publish the app, don't publish the `.sln` file (because if you do, the SelfContained property will be set for the Client project - since it is a "Global" property). Instead publish the `Server.csproj`. Modifying your command above:

```bash
dotnet publish ./BlazorWasmHosted/Server --runtime linux-x64 --no-self-contained
```

That allowed this scenario to work for me.

**Comment by yugabe** [comment]
Thank you for the workaround, @eerhardt. 

It seems reasonable to take this information and supply a fix at the SDK level?
I guess, from what you said, that when building a Blazor application specifically for the browser(-wasm) target, it should always be self-contained, so the build shouldn't just break, it should ignore the switch altogether.

Although you are right in that I was publishing the solution directly in my example, it was an oversight on my part -- as in my production CI pipeline I build the Server project directly, and it produces a different result.

That said, is this issue correctly categorized here, or should it go the github.com/dotnet/aspnetcore repo?

**Comment by agocke** [comment]
I'm not an expert on Blazor so take all this with a grain of salt but,

If SelfContained=false is being passed to a project with mandatory trimming, I'd expect the trimming targets to error about that. Maybe something else could fix things up at a higher level, but the input being provided to trimming is legitimately nonsensical and indicates some kind of configuration bug upstream, and I'd generally like an error about that.

**Comment by yugabe** [comment]
@agocke This I know to be a bit different (accidentally). In a Blazor WASM hosted app (with prerendering), the client bits needs to be built twice: once for running in the browser and once for running on the server. 

When building for the browser (the resulting assemblies will be served as static files from the server's wwwroot), the app obviously needs to be self-contained as there is no .NET runtime available in the browser. It is trimmed fully by default from .NET 7 onwards too. 

When being built for the server, it's just another assembly to be built for the target runtime (linux-x64 in my case), so it should be handled as any other library project.

The issue I reckon is that somehow these two methods to build the client app need to be configured differently, and the logic that decides how that should be done is possibly flawed. So, the SelfContained flag should always be ignored/should default to true when building the Client for the browser, and should be ambient/configurable when not. The problem is also that *both* can and do happen in the same build cycle when doing a full build of a Blazor Hosted WASM Server app.

**Comment by javiercn** [comment]
> ```
>   <PropertyGroup>
>     <StaticWebAssetsAdditionalPublishPropertiesToRemove>$(StaticWebAssetsAdditionalPublishPropertiesToRemove);SelfContained</StaticWebAssetsAdditionalPublishPropertiesToRemove>
>   </PropertyGroup>
> ```

We could potentially do this in our targets.

The issue is more or less what @yugabe mentioned. There is an invocation for the client project to get the dlls from referenced projects and there is a separate invocation we do to get the files that need to go into the wwwroot folder. I believe we already do work to ensure that on blazor webassembly the rid is always wasm [here](https://github.com/dotnet/sdk/blob/main/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props#L15) and we remove other properties [here](https://github.com/dotnet/sdk/blob/main/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props#L40)

**Comment by muayyat** [comment]
Any solution on this yet? Using 7.0.2 and still facing same problem

**Comment by agocke** [comment]
I think the fix here has to be in ASP.NET. Moving to that repo.
